kani-compiler/src/kani_middle/transform/contracts.rs

```diff
@@ -232,8 +228,297 @@ impl AnyModifiesPass {
+                            target.unwrap(),
+                        ));
+                    }
+                }
+                None
+            })
+            .unwrap();
+
+        let span = mode_call.span(new_body.blocks());
+        let mode_const = new_body.new_uint_operand(mode as _, UintTy::U8, span);
+        new_body.assign_to(
+            ret.clone(),
+            Rvalue::Use(mode_const),
+            &mut mode_call,
+            InsertPosition::Before,
+        );
+        new_body.replace_terminator(
+            &mode_call,
+            Terminator { kind: TerminatorKind::Goto { target }, span },
+        );
+
+        new_body.into()
+    }
+
+    /// Return which contract mode to use for this function if any.
+    fn contract_mode(&self, tcx: TyCtxt, fn_def: FnDef) -> Option<ContractMode> {
+        let kani_attributes = KaniAttributes::for_def_id(tcx, fn_def.def_id());
+        kani_attributes.has_contract().then(|| {
+            let fn_def_id = rustc_internal::internal(tcx, fn_def.def_id());
+            if self.check_fn == Some(fn_def_id) {
+                if kani_attributes.has_recursion() {
+                    ContractMode::RecursiveCheck
+                } else {
+                    ContractMode::SimpleCheck
+                }
+            } else if self.replace_fns.contains(&fn_def_id) {
+                ContractMode::Replace
+            } else {
+                ContractMode::Original
+            }
+        })
+    }
+
+    /// Select any unused closure for body deletion.
+    fn mark_unused(&mut self, tcx: TyCtxt, fn_def: FnDef, body: &Body, mode: ContractMode) {
+        let contract =
+            KaniAttributes::for_def_id(tcx, fn_def.def_id()).contract_attributes().unwrap();
+        let recursion_closure = find_closure(tcx, fn_def, &body, contract.recursion_check.as_str());
+        let check_closure = find_closure(tcx, fn_def, &body, contract.checked_with.as_str());
+        let replace_closure = find_closure(tcx, fn_def, &body, contract.replaced_with.as_str());
+        match mode {
+            ContractMode::Original => {
+                // No contract instrumentation needed. Add all closures to the list of unused.
+                self.unused_closures.insert(recursion_closure);
+                self.unused_closures.insert(check_closure);
+                self.unused_closures.insert(replace_closure);
+            }
+            ContractMode::RecursiveCheck => {
+                self.unused_closures.insert(replace_closure);
+                self.unused_closures.insert(check_closure);
+            }
+            ContractMode::SimpleCheck => {
+                self.unused_closures.insert(replace_closure);
+                self.unused_closures.insert(recursion_closure);
+            }
+            ContractMode::Replace => {
+                self.unused_closures.insert(recursion_closure);
+                self.unused_closures.insert(check_closure);
+            }
```

Comment:
If the idea is to completely replace the macros with closures, the way we need to handle modifies clauses doesn't impact at all here? We can still generate the inner function with macros?

Reply:
I'm assuming you mean replace functions with closures. The modifies clauses is not important for the transformation because its logic is not directly reachable from the original function.

We don't need to delete the body, since it will be unreachable once we delete the body of the `check_closure` and `recursion_closure`.